File: src/components/study/isc-12/topics/computational-complexity-big-o/topic16_files/RecursiveFactorial.java
/**
 * Recursive factorial function.
 * Recurrence: T(n) = T(n-1) + O(1), T(0) = O(1)
 * Time: O(n), Space: O(n) (recursion stack depth = n)
 */
public class RecursiveFactorial {
    public static void main(String[] args) {
        int n = 5;
        long result = factorial(n);
        System.out.println(n + "! = " + result);

        // Show call count and depth
        System.out.println("Number of calls: " + (n + 1));
        System.out.println("Recursion depth: " + n);
        System.out.println("Time: O(n), Space: O(n)");
    }

    public static long factorial(int n) {
        // Base case
        if (n <= 1) return 1;

        // Recursive case: T(n) = T(n-1) + O(1)
        return n * factorial(n - 1);
    }
}

File: src/components/study/c-language-old/topics/c-functions-basics/topic45_files/recursive_factorial.c
// recursive_factorial.c – Recursive factorial (linear recursion)
#include <stdio.h>

long long factorial(int n) {
    if (n < 0) return -1;
    if (n == 0 || n == 1) return 1;
    return n * factorial(n - 1);  // Recursive call
}

int main() {
    int n = 5;
    printf("%d! = %lld\n", n, factorial(n));
    return 0;
}